modules/table/src/test/java/org/apache/ignite/internal/table/distributed/raft/snapshot/incoming/IncomingSnapshotCopierTest.java

```diff
@@ -0,0 +1,322 @@
+                            .meta(RAFT_MSG_FACTORY.snapshotMeta().lastIncludedIndex(outgoingPartitionStorage.lastAppliedIndex()).build())
+                            .build()
+            );
+        });
+
+        when(messagingService.invoke(eq(clusterNode), any(SnapshotMvDataRequest.class), anyLong())).then(answer -> {
+            SnapshotMvDataRequest snapshotMvDataRequest = answer.getArgument(1);
+
+            assertEquals(snapshotId, snapshotMvDataRequest.id());
+
+            List<ResponseEntry> responseEntries = createResponseEntries(outgoingPartitionStorage, rowIds);
+
+            assertThat(responseEntries, not(empty()));
+
+            return completedFuture(TABLE_MSG_FACTORY.snapshotMvDataResponse().rows(responseEntries).finish(true).build());
+        });
+
+        when(outgoingSnapshotsManager.messagingService()).thenReturn(messagingService);
+
+        return new PartitionSnapshotStorage(
+                topologyService,
+                outgoingSnapshotsManager,
+                SnapshotUri.toStringUri(snapshotId, NODE_NAME),
+                mock(RaftOptions.class),
+                new MvPartitionAccess(incomingTableStorage, TEST_PARTITION),
+                mock(SnapshotMeta.class),
+                executorService
+        );
+    }
+
+    private static List<RowId> prepareOutgoingPartitionStorage(MvPartitionStorage partitionStorage) {
+        List<RowId> rowIds = new ArrayList<>();
+
+        partitionStorage.runConsistently(() -> {
+            // Writes committed version.
+            RowId rowId0 = new RowId(TEST_PARTITION);
+            RowId rowId1 = new RowId(TEST_PARTITION);
+
+            partitionStorage.addWriteCommitted(rowId0, createBinaryRow("k0", "v0"), HYBRID_CLOCK.now());
+            partitionStorage.addWriteCommitted(rowId1, createBinaryRow("k1", "v1"), HYBRID_CLOCK.now());
+
+            RowId rowId2 = new RowId(TEST_PARTITION);
+
+            partitionStorage.addWriteCommitted(rowId2, createBinaryRow("k20", "v20"), HYBRID_CLOCK.now());
+            partitionStorage.addWriteCommitted(rowId2, createBinaryRow("k21", "v21"), HYBRID_CLOCK.now());
+
+            // Writes an intent to write (uncommitted version).
+            partitionStorage.addWrite(rowId2, createBinaryRow("k22", "v22"), UUID.randomUUID(), UUID.randomUUID(), TEST_PARTITION);
+
+            RowId rowId3 = new RowId(TEST_PARTITION);
+
+            partitionStorage.addWrite(
+                    rowId3,
+                    createBinaryRow("k3", "v3"),
+                    UUID.randomUUID(),
+                    UUID.randomUUID(),
+                    TEST_PARTITION
+            );
+
+            rowIds.addAll(Set.of(rowId0, rowId1, rowId2, rowId3));
+
+            partitionStorage.lastAppliedIndex(rowIds.size());
+
+            return null;
+        });
+
+        return rowIds;
+    }
+
+    private static List<ResponseEntry> createResponseEntries(MvPartitionStorage partitionStorage, List<RowId> rowIds) {
```

Comment:
Again, having some comments here and there would be nice

Reply:
Tried to fix it.